src/mechanics/cell/CellData.h

```diff
@@ -14,26 +14,36 @@ namespace NuTo
 class CellData
 {
 public:
-    CellData(const DofContainer<ElementInterface*>& elements)
+    CellData(const ElementCollection& elements)
         : mElements(elements)
     {
     }
 
+    NodeValues GetCoordinates() const
+    {
+        return mElements.CoordinateElement().ExtractNodeValues();
+    }
+
     NodeValues GetNodeValues(const DofType& dofType) const
     {
         NodeValues& nodeValues = mNodeValues[dofType];
         if (nodeValues.size() == 0)
         {
+            /* The mutable member mNodeValues may be accessed by multiple threads
+             * at once - in an OpenMP context.
+             * Concurrent read access is no problem. This if branch, however,
+             * performs write access. No concurrency here.
+             */
 #ifdef _OPENMP
 #pragma omp critical
 #endif
-            nodeValues = mElements[dofType]->ExtractNodeValues();
+            nodeValues = mElements.DofElement(dofType).ExtractNodeValues();
```

Comment:
Should we really add an omp critical here. Each cell and thus each cell_data should be accessed only by a single thread, thus this should not be an issue at all.